framework/include/base/ResidualObject.h

```diff
@@ -0,0 +1,118 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "MooseObject.h"
+#include "SetupInterface.h"
+#include "CoupleableMooseVariableDependencyIntermediateInterface.h"
+#include "FunctionInterface.h"
+#include "UserObjectInterface.h"
+#include "TransientInterface.h"
+#include "PostprocessorInterface.h"
+#include "VectorPostprocessorInterface.h"
+#include "RandomInterface.h"
+#include "Restartable.h"
+#include "MeshChangedInterface.h"
+#include "TaggingInterface.h"
+
+class MooseMesh;
+class SubProblem;
+class KernelBase;
+class Assembly;
+template <typename>
+class MooseVariableFE;
+typedef MooseVariableFE<Real> MooseVariable;
+typedef MooseVariableFE<VectorValue<Real>> VectorMooseVariable;
+typedef MooseVariableFE<RealEigenVector> ArrayMooseVariable;
+
+template <>
+InputParameters validParams<KernelBase>();
+
+/**
+ * This is the common base class for objects that give residual contributions.
+ */
+class ResidualObject : public MooseObject,
+                       public SetupInterface,
+                       public FunctionInterface,
+                       public UserObjectInterface,
+                       public TransientInterface,
+                       public PostprocessorInterface,
+                       public VectorPostprocessorInterface,
+                       public RandomInterface,
+                       public Restartable,
+                       public MeshChangedInterface,
+                       public TaggingInterface
+{
+public:
+  static InputParameters validParams();
+
+  ResidualObject(const InputParameters & parameters);
+
+  /// Compute this Kernel's contribution to the residual
+  virtual void computeResidual() = 0;
+
+  /// Compute this Kernel's contribution to the diagonal Jacobian entries
+  virtual void computeJacobian() = 0;
+
+  /// Computes d-residual / d-jvar... storing the result in Ke.
+  virtual void computeOffDiagJacobian(MooseVariableFEBase & /*jvar*/) {}
```

Comment:
It might be even better to use `MooseVariableBase` to cover all cases

Reply:
That will be too aggressive.

Reply:
We want the parameter here to be `unsigned int jvar` instead of `MooseVariable.*`. Having this parameter as `MooseVariable.*` can more easily lead to errors like the one you encountered in the displaced integrated boundary condition test. This will require modifying `ComputeFullJacobianThread::computeFaceJacobian` but that's ok